app/src/lib/stores/commit-status-store.ts

```diff
@@ -0,0 +1,336 @@
+    if (this.backgroundRefreshHandle === null) {
+      this.backgroundRefreshHandle = window.setInterval(
+        () => this.queueRefresh(),
+        BackgroundRefreshInterval
+      )
+      this.queueRefresh()
+    }
+  }
+
+  /**
+   * Called to ensure that background refreshing is stopped. The intention
+   * is for background refreshing to be active while the application is
+   * focused.
+   *
+   * Remarks: this method will do nothing if background fetching is
+   *          not currently active.
+   */
+  public stopBackgroundRefresh() {
+    if (this.backgroundRefreshHandle !== null) {
+      window.clearInterval(this.backgroundRefreshHandle)
+      this.backgroundRefreshHandle = null
+    }
+  }
+
+  private queueRefresh() {
+    if (!this.refreshQueued) {
+      this.refreshQueued = true
+      setImmediate(() => {
+        this.refreshQueued = false
+        this.refreshEligibleSubscriptions()
+      })
+    }
+  }
+
+  /**
+   * Looks through all active commit status subscriptions and
+   * figure out which, if any, needs to be refreshed from the
+   * API.
+   */
+  private refreshEligibleSubscriptions() {
+    for (const key of this.subscriptions.keys()) {
+      // Is it already being worked on?
+      if (this.queue.has(key)) {
+        continue
+      }
+
+      const entry = this.cache.get(key)
+
+      if (entry && !entryIsEligibleForRefresh(entry)) {
+        continue
+      }
+
+      this.limit(() => this.refreshSubscription(key))
+        .catch(e => log.error('Failed refreshing commit status', e))
+        .then(() => this.queue.delete(key))
+
+      this.queue.add(key)
+    }
+  }
+
+  private async refreshSubscription(key: string) {
+    // Make sure it's still a valid subscription that
+    // someone might care about before fetching
+    const subscription = this.subscriptions.get(key)
+
+    if (subscription === undefined) {
+      return
+    }
+
+    const { endpoint, owner, name, ref } = subscription
+    const account = this.accounts.find(a => a.endpoint === endpoint)
+
+    if (account === undefined) {
+      return
+    }
+
+    try {
+      const api = API.fromAccount(account)
+      const status = await api.fetchCombinedRefStatus(owner, name, ref)
+
+      this.cache.set(key, { status, fetchedAt: new Date() })
+      subscription.callbacks.forEach(cb => cb(status))
+    } catch (err) {
+      // Okay, so we failed retrieving the status for one reason or another.
+      // That's a bummer, but we still need to put something in the cache
+      // or else we'll consider this subscription eligible for refresh
+      // from here on until we succeed in fetching. By putting a blank
+      // cache entry (or potentially reusing the last entry) in and not
+      // notifying subscribers we ensure they keep their current status
+      // if they have one and that we attempt to fetch it again on the same
+      // schedule as the others.
+      log.debug(`Failed retrieving commit status for ref ${ref}`, err)
```

Comment:
🤔 because we removed the error handling inside `fetchCombinedRefStatus` so it will bubble up we also lost the context of which `owner/repo` associated with the failing request. Is that worth tracking to make supporting these changes better? This is also only visible in the Dev Tools console, and not logged to disk, which I think is fine.

Reply:
It's only visible in devtools [_while you're running in dev mode_](https://github.com/desktop/desktop/blob/d5a938749512b5086c853a9c5640c0ef7992efb8/app/src/main-process/log.ts#L54-L56) if I'm understanding things correctly and in that case I'm assuming that the ref is enough context to understand what's going on but I guess there's no harm in including owner & name either.